dvc/repo/ls.py

```diff
@@ -66,16 +66,28 @@ def onerror(exc):
 
     ret = {}
     for info in infos:
-        metadata = fs.metadata(info)
-        if metadata.output_exists or not dvc_only:
-            path = (
-                fs.path.name(fs_path)
-                if fs_path == info
-                else fs.path.relpath(info, fs_path)
-            )
-            ret[path] = {
-                "isout": metadata.is_output,
-                "isdir": metadata.isdir,
-                "isexec": metadata.is_exec,
-            }
+        path = (
+            fs.path.name(fs_path)
+            if fs_path == info
+            else fs.path.relpath(info, fs_path)
+        )
+
+        try:
+            metadata = fs.metadata(info)
+            if metadata.output_exists or not dvc_only:
+                ret[path] = {
+                    "isout": metadata.is_output,
+                    "isdir": metadata.isdir,
+                    "isexec": metadata.is_exec,
+                }
+
+        except FileNotFoundError:
+            # if walk found it but there is no metadata, its broken symlink
+            if not dvc_only:
+                ret[path] = {
+                    "isout": False,
+                    "isdir": False,
+                    "isexec": False,
+                }
+
```

Comment:
Note that you are ignoring FileNotFoundError if `dvc_only`, which could be a dangerous bug. Also duplicating code, so how about smth like:

```
        try:
            metadata = fs.metadata(info)
        except FileNotFoundError:
            if dvc_only:
                raise
                
            metadata = Metadata()
            
        if metadata.output_exists or not dvc_only:
            path = (
                fs.path.name(fs_path)
                if fs_path == info
                else fs.path.relpath(info, fs_path)
            )
            ret[path] = {
                "isout": metadata.is_output,
                "isdir": metadata.isdir,
                "isexec": metadata.is_exec,
            }
```